design/MultipleQueues.md

```diff
@@ -0,0 +1,24 @@
+# Multiple Queues
+
+Queues are created at Device creation time, and are exposed via `sequence<Queue> Device.queues`.
+There are multiple types of queues, denoted by the `.type` bitfield on the Queue object.
+Queues can support `graphics` operations, `compute` operations, both, or neither.
+"Neither" Queues can only execute copy commands.
+If there are any queues that support graphics, there is at least one queue that supports
+both compute and graphics.
+The most capable queue is always first in the `Device.queues` list.
+(It will either be a graphics+compute queue, or compute queue)
+Devices must be able to support at least a compute queue, or device creation will fail.
+
+## Synchronization
+
+Queues can have fences inserted into them, and any queue can wait on that fence to be complete.
+Synchronizing usage of resources across queues should be done by telling a queue to wait until
+a fence is complete, followed by submitting command buffers that depend on that fence.
+
+## Further work
+
+- Transfering resources between queues, particularly for VK_SHARING_MODE_EXCLUSIVE, as opposed to _CONCURRENT.
```

Comment:
D3D12's [D3D12_RESOURCE_FLAG_ALLOW_SIMULTANEOUS_ACCESS](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/ne-d3d12-d3d12_resource_flags) flag is relevant here.

Simultaneous access guarentees that a resource is in a layout that can be read from any queue type. Hence, it allows you to mix SRV and UAV states across different queues running at the same time. If you know you're not going be mixing these for the same resource, you can leave the flag off.

Regardless of the state fo the flag, undefined behavior will result if you write to a region of a resource that is being read from an operation on a different queue or the same queue.

Reply:
+1, we need to figure out how we are going to avoid write-write and read-write hazards.

Reply:
In vulkan only one queue family can be owner of a resource, AFAIK

Reply:
Not with `VK_SHARING_MODE_CONCURRENT`

Reply:
"Ranges of buffers and image subresources of image objects created using VK_SHARING_MODE_CONCURRENT must only be accessed by queues from the queue families specified through the queueFamilyIndexCount and pQueueFamilyIndices members of the corresponding create info structures."

You also need to specify up-front which families you will share with when creating the objects.

Also the performance impact
"
(info) "VK_SHARING_MODE_CONCURRENT may result in lower performance access to the buffer or image than VK_SHARING_MODE_EXCLUSIVE"
"
https://www.reddit.com/r/vulkan/comments/7m6h0x/just_using_vk_sharing_mode_concurrent/
**SachaWillems** - "You'll notice differences on lower spec devices, esp. on mobile. So if you want to support mobile too, just using concurrent all the time is not a good idea, as it also forces you to always specify (up front) what queues will access the image."